### PR TITLE
 Logger

### PR DESCRIPTION
**Pull request by cliveseldon**
Adds inference logging as defined [here](https://docs.google.com/document/d/1dCzOcCmeOEAaJwDIB2F6GkrRoIicRDGo5ocxZD9vzFs/edit#heading=h.ylpx1kb2u1hv).

Key parts:

 * Updates control plane with optional inference_logger that can be applied to any predictor, transformer or explainer.
 * Adds inference-logger side-car to container to asynchronously send CloudEvents for request, response or both to user configured URL
 * Adds samples in `docs/samples/inference-logger`
 * Vendor folder is updated with CloudEvents libraries

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/467)

### CODE REVIEW

**Comment by ellis-bigelow** [comment]
Excited to see this. Highly recommend we separate this into 4-5 changes of a couple hundred lines each max. Perhaps:

1. go types/interface
2. dependencies
3. docs
4. logger

**Comment by cliveseldon** [comment]
@ellis-bigelow  I have added #472 and #473 
When these are reviewed and merged I can add the others.

**Comment by cliveseldon** [comment]
These are the remaining changes:

  * Reconcile
  * Demos

If you could review @ellis-bigelow

File: tools/tf2openapi/generated/framework/remote_fused_graph_execute_info.pb.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ type RemoteFusedGraphExecuteInfo struct {
 	GraphOutputNodeName []string `protobuf:"bytes,3,rep,name=graph_output_node_name,json=graphOutputNodeName,proto3" json:"graph_output_node_name,omitempty"`
 	// Executor's name
 	ExecutorName string `protobuf:"bytes,4,opt,name=executor_name,json=executorName,proto3" json:"executor_name,omitempty"`
-	// Optional: Parameters given to the executor
+	// Optional: Parameters given to the inference-logger
```
Comment (ellis-bigelow) [inline tools/tf2openapi/generated/framework/remote_fused_graph_execute_info.pb.go]: lol! something wrong here?

Comment (cliveseldon) [reply tools/tf2openapi/generated/framework/remote_fused_graph_execute_info.pb.go]: Not sure how that got there. Anyway, I have rerun the Makefile to regenerate the files.

File: pkg/webhook/admission/pod/logger_injector.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+type LoggerConfig struct {
+	Image string `json:"image"`
+}
+
+type LoggerInjector struct {
+	config *LoggerConfig
+}
+
+func (il *LoggerInjector) InjectLogger(pod *v1.Pod) error {
```
Comment (yuzisun) [inline pkg/webhook/admission/pod/logger_injector.go]: should we do this on podspec instead of injecting in mutator? knative does not support init containers but sidecar is fine right?

Comment (cliveseldon) [reply pkg/webhook/admission/pod/logger_injector.go]: This not allowed at present.  See:
  * https://github.com/knative/serving/issues/3384
  *  https://github.com/knative/serving/issues/5822

File: docs/samples/logger/basic/sklearn-logging.yaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+      minReplicas: 1      
+      logger:
+        url: http://message-dumper.default/
+        mode: all
+      sklearn:
+        storageUri: "gs://kfserving-samples/models/sklearn/iris"
+        resources:
+          requests:
+            cpu: 0.1
+
```
Comment (ellis-bigelow) [inline docs/samples/logger/basic/sklearn-logging.yaml]: nit extra line

File: docs/samples/logger/basic/sklearn-logging.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+spec:
+  default:
+    predictor:
+      minReplicas: 1      
+      logger:
+        url: http://message-dumper.default/
+        mode: all
+      sklearn:
+        storageUri: "gs://kfserving-samples/models/sklearn/iris"
+        resources:
```
Comment (ellis-bigelow) [inline docs/samples/logger/basic/sklearn-logging.yaml]: Why not just let this default? 0.1 is quite restrictive

File: docs/samples/sklearn/sklearn.yaml [hunk 5]
```diff
@@ -7,3 +7,7 @@ spec:
     predictor:
       sklearn:
         storageUri: "gs://kfserving-samples/models/sklearn/iris"
+        resources:
```
Comment (ellis-bigelow) [inline docs/samples/sklearn/sklearn.yaml]: Why set such small defaults?

File: docs/samples/logger/knative-eventing/sklearn-logging.yaml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  name: "sklearn-iris"
+spec:
+  default:
+    predictor:
+      minReplicas: 1
+      logger:
+        mode: all
+      sklearn:
+        storageUri: "gs://kfserving-samples/models/sklearn/iris"
+        resources:
```
Comment (ellis-bigelow) [inline docs/samples/logger/knative-eventing/sklearn-logging.yaml]: likewise.

File: docs/samples/logger/knative-eventing/README.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+```
+apiVersion: "serving.kubeflow.org/v1alpha2"
+kind: "InferenceService"
+metadata:
+  name: "sklearn-iris"
+spec:
+  default:
+    predictor:
+      logger:
+        logType: all
```
Comment (ellis-bigelow) [inline docs/samples/logger/knative-eventing/README.md]: outdated?

File: docs/samples/logger/knative-eventing/README.md [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+      sklearn:
+        storageUri: "gs://kfserving-samples/models/sklearn/iris"
+        resources:
```
Comment (ellis-bigelow) [inline docs/samples/logger/knative-eventing/README.md]: likewise

File: pkg/apis/serving/v1alpha2/explainer.go [hunk 9]
```diff
@@ -67,6 +67,7 @@ func (e *ExplainerSpec) Validate(config *InferenceServicesConfig) error {
 		validateStorageURI(e.GetStorageUri()),
 		validateReplicas(e.MinReplicas, e.MaxReplicas),
 		validateResourceRequirements(explainer.GetResourceRequirements()),
+		validate_logger(e.Logger),
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha2/explainer.go]: nit: casing

File: pkg/apis/serving/v1alpha2/explainer.go [hunk 10]
```diff
@@ -42,12 +42,12 @@ func (e *ExplainerSpec) GetStorageUri() string {
 	return explainer.GetStorageUri()
 }
 
-func (e *ExplainerSpec) CreateExplainerContainer(modelName string, predictorHost string, config *InferenceServicesConfig) *v1.Container {
+func (e *ExplainerSpec) CreateExplainerContainer(modelName string, predictorHost string, config *InferenceServicesConfig, hasLogging bool) *v1.Container {
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha2/explainer.go]: nit: this arg should come before the injected config.

File: pkg/apis/serving/v1alpha2/framework_pytorch.go [hunk 11]
```diff
@@ -47,6 +47,7 @@ func (s *PyTorchSpec) GetContainer(modelName string, config *InferenceServicesCo
 			"--model_name=" + modelName,
 			"--model_class_name=" + s.ModelClassName,
 			"--model_dir=" + constants.DefaultModelLocalMountPath,
+			"--http_port=" + constants.GetInferenceServiceHttpPort(hasLogging),
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha2/framework_pytorch.go]: It's a little bit crufty to wire through the logging component into all of our concepts. What do you think about using a more "sidecar"-y approach.

We could have a mutator that injects a second container into the workload and remaps the ports using: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#containerport-v1-core.

This way, from the perspective of the predictor, it's still serving on the same port. From the perspective of the pod sepc, we can wire up our logger to sit where knative thinks it should be (i.e. 8080) and call out to the user containers new port (i.e. 8081?). All of these details can be fully encapsulated inside the sidecar injector.

WDYT?

File: pkg/webhook/admission/pod/logger_injector.go [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+	}
+
+	logMode, ok := pod.ObjectMeta.Annotations[constants.LoggerModeInternalAnnotationKey]
+	if !ok {
+		logMode = string(v1alpha2.LogAll)
+	}
+
+	modelId, _ := pod.ObjectMeta.Labels[PodKnativeServiceLabel]
+
+	// Dont inject if Contianer already injected
```
Comment (ellis-bigelow) [inline pkg/webhook/admission/pod/logger_injector.go]: nit: Don't, Container

File: pkg/webhook/admission/pod/logger_injector.go [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+		if strings.Compare(container.Name, LoggerContainerName) == 0 {
+			return nil
+		}
+	}
+
+	loggerContainer := &v1.Container{
+		Name:  LoggerContainerName,
+		Image: il.config.Image,
+		Args: []string{
+			"--log-url",
```
Comment (ellis-bigelow) [inline pkg/webhook/admission/pod/logger_injector.go]: Can we make these constants?

File: pkg/webhook/admission/pod/logger_injector.go [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+			return loggerConfig, err
+		}
+	}
+
+	return loggerConfig, nil
+}
+
+func (il *LoggerInjector) InjectLogger(pod *v1.Pod) error {
+	// Only inject if the required annotations are set
+	_, ok := pod.ObjectMeta.Annotations[constants.LoggerInternalAnnotationKey]
```
Comment (ellis-bigelow) [inline pkg/webhook/admission/pod/logger_injector.go]: Instead of passing all of these through as annotations, can we simply read the pod's labels to get the inferenceservice and get the logging config from there? For this approach to be clean, we'd need the InferenceService's name, the component, and a label. We could then use our pkg/types code to have something like logger.GetContainerSpec the same way we do w/ other components. 

I'm trying to keep as much logic out of the webhook as possible, and encapsulate it within our types so that if we move away from a webhook approach, it doesn't take a big refactor.

WDYT?

File: docs/samples/logger/basic/sklearn-logging.yaml (hunk 4 above)
Comment (cliveseldon) [reply docs/samples/logger/basic/sklearn-logging.yaml]: ok. Just as these are demos can easily use up cluster capacity. But probably cleaner.

File: pkg/apis/serving/v1alpha2/framework_pytorch.go (hunk 11 above)
Comment (cliveseldon) [reply pkg/apis/serving/v1alpha2/framework_pytorch.go]: I don't think this is possible as all containers share a network on the pod so ports can't be conflicting. `Number of port to expose on the pod's IP address` - note reference to Pod here.

File: pkg/webhook/admission/pod/logger_injector.go (hunk 14 above)
Comment (cliveseldon) [reply pkg/webhook/admission/pod/logger_injector.go]: I think I would like to punt on this as it will add quite alot of complication to the mutator and general tests. Can we do this as an improvement later as we develop the use of the storage and logger mutators?

**Comment by ellis-bigelow** [comment]
/lgtm
/approve

Looks like it needs a rebase

File: pkg/apis/serving/v1alpha2/framework_pytorch.go (hunk 11 above)
Comment (pugangxa) [reply pkg/apis/serving/v1alpha2/framework_pytorch.go]: Knative default to use 8080 but can specify one port by the containerport in spec.
https://github.com/knative/serving/blob/24a4448a9d520e02aef1ebc3d041826d2b3819ff/pkg/reconciler/revision/resources/deploy.go#L174-L181

So I think we need not do it in injector or remap it. We can just make the logger always listen on 8081 and the predictor always on 8080, but check the hasLogger condition and if true add this when we reconcile knative service
```
          ports:
            - name: http1
              containerPort: 8081
```
In this way although we still need the haslogger condition but we need not redefine the port, and also we need not do it in every predictor/transformer/explainer, but can just add this on the top level of knative service resource.

Comment (cliveseldon) [reply pkg/apis/serving/v1alpha2/framework_pytorch.go]: This seems like a good idea to clean the integration. As we are not covering Custom containers.

Comment (pugangxa) [reply pkg/apis/serving/v1alpha2/framework_pytorch.go]: I have implemented this idea in PR #592  , please take a look.
